### PR TITLE
[RayService] Add e2e tests

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

4 e2e tests for RayService:
- deploy 2 applications
- deploy then execute in place update
- deploy then execute zero downtime rollout
- autoscaling

The e2e tests should be run with the following command:
```
RAY_IMAGE=rayproject/ray:2.5.0 OPERATOR_IMAGE=controller:latest pytest -vs tests/test_sample_rayservice_yamls.py --log-cli-level=INFO
```

To run a specific test in `test_sample_rayservice_yamls.py`, specify the test name with `-k`, e.g:
```
RAY_IMAGE=rayproject/ray:2.5.0 OPERATOR_IMAGE=controller:latest pytest -vs tests/test_sample_rayservice_yamls.py -k test_service_autoscaling --log-cli-level=INFO
```

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/config/samples/ray-service.autoscaler.yaml [hunk 1]
```diff
@@ -9,72 +9,56 @@ metadata:
...
-        routePrefix: "/"
-        rayActorOptions:
-          numCpus: 0.1
+  serveConfigV2: |
+    applications:
+      - name: app1
+        import_path: blocked.app
+        route_prefix: /
+        runtime_env:
+          working_dir: "https://github.com/zcin/sample-serve-applications/archive/refs/heads/main.zip"
```
Comment (shrekris-anyscale) [inline ray-operator/config/samples/ray-service.autoscaler.yaml:18]: Can you move this application to [ray-project](https://github.com/ray-project) instead of your personal GitHub? I'd suggest either making a new repo there or using [this one](https://github.com/ray-project/serve_workloads).

Comment (shrekris-anyscale) [reply ray-operator/config/samples/ray-service.autoscaler.yaml:18]: And could you use a specific commit hash instead of linking to `main`? That prevents the test from failing if someone pushes a broken change to the repo.

File: ray-operator/config/samples/ray-service.autoscaler.yaml [hunk 2]
```diff
@@ -9,72 +9,56 @@ metadata:
...
+              upscale_delay_s: 0
+            graceful_shutdown_timeout_s: 1
+            max_concurrent_queries: 1000
+            ray_actor_options:
+              num_cpus: 0.5
+      - name: signal
+        import_path: signaling.app
+        route_prefix: /signal
+        runtime_env:
+          working_dir: "https://github.com/zcin/sample-serve-applications/archive/refs/heads/main.zip"
```
Comment (shrekris-anyscale) [inline ray-operator/config/samples/ray-service.autoscaler.yaml:36]: Same recommendations here about moving the code to the `ray-project` GitHub.

File: tests/framework/prototype.py [hunk 3]
```diff
@@ -186,7 +195,7 @@ def check_rule_sets(self):
 
     def clean_up(self):
         """Cleanup the CR."""
-        raise NotImplementedError
+        pass
```
Comment (shrekris-anyscale) [inline tests/framework/prototype.py:198]: Do we need this function if it just runs `pass`?

File: tests/test_sample_rayservice_yamls.py [hunk 4]
```diff
@@ -1,55 +1,338 @@
...
-
-    # Execute all tests
-    runner = unittest.TextTestRunner()
-    test_result = runner.run(test_cases)
-
-    # Without this line, the exit code will always be 0.
-    assert test_result.wasSuccessful()
+NAMESPACE = 'default'
+DEFAULT_IMAGE_DICT = {
+    CONST.RAY_IMAGE_KEY: os.getenv('RAY_IMAGE', default='rayproject/ray:2.5.0'),
```
Comment (shrekris-anyscale) [inline tests/test_sample_rayservice_yamls.py:38]: Should we use the `nightly` instead of 2.5.0 here?

File: tests/test_sample_rayservice_yamls.py [hunk 5]
```diff
@@ -1,55 +1,338 @@
...
+                return
+
+            time.sleep(1)
+
+        logger.info(f"RayServiceAddCREvent wait() failed to converge in {self.timeout}s.")
+        logger.info(
+            f"expected_head_pods: {expected_head_pods}, "
+            f"expected_worker_pods: {expected_worker_pods}"
+        )
+        show_cluster_info(self.namespace)
```
Comment (shrekris-anyscale) [inline tests/test_sample_rayservice_yamls.py:85]: Does this log the number of running head pods and worker pods? If not, could you log those?

File: tests/framework/prototype.py (hunk 3 above)
Comment (zcin) [reply tests/framework/prototype.py:198]: Hmm, I haven't looked into the other tests that use CREvent, but I think they all call clean_up(). I changed it to `pass` because the new CREvents I wrote don't run clean_up.

Comment (zcin) [reply tests/framework/prototype.py:198]: Actually, I can revert this change.

File: ray-operator/config/samples/ray-service.autoscaler.yaml (hunk 1 above)
Comment (zcin) [reply ray-operator/config/samples/ray-service.autoscaler.yaml:18]: Thanks for the suggestion! I've pushed it to the repo you linked.

File: tests/test_sample_rayservice_yamls.py (hunk 4 above)
Comment (zcin) [reply tests/test_sample_rayservice_yamls.py:38]: Hmm not sure what the convention is, but the Kuberay and Anyscale tests I've seen are run against the latest version of Ray.

File: tests/test_sample_rayservice_yamls.py (hunk 5 above)
Comment (zcin) [reply tests/test_sample_rayservice_yamls.py:85]: Logged!

File: tests/test_sample_rayservice_yamls.py (hunk 4 above)
Comment (shrekris-anyscale) [reply tests/test_sample_rayservice_yamls.py:38]: Got it– in that case, let's leave the version as is.

File: ray-operator/config/samples/ray-service.autoscaler.yaml (hunk 1 above)
Comment (shrekris-anyscale) [reply ray-operator/config/samples/ray-service.autoscaler.yaml]: Great, thanks!

**Review by shrekris-anyscale — APPROVED** [review]
Thanks for addressing all my comments! The change looks good to me.

File: ray-operator/config/samples/ray-service.autoscaler.yaml [hunk 6]
```diff
@@ -9,72 +9,56 @@ metadata:
...
-        requests:
-          cpu: "1"
-          memory: "1000Mi"
     ######################headGroupSpecs#################################
     headGroupSpec:
       # The `rayStartParams` are used to configure the `ray start` command.
       # See https://github.com/ray-project/kuberay/blob/master/docs/guidance/rayStartParams.md for the default settings of `rayStartParams` in KubeRay.
       # See https://docs.ray.io/en/latest/cluster/cli.html#ray-start for all available options in `rayStartParams`.
-      rayStartParams: {"num-cpus": "0"}
+      rayStartParams:
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.autoscaler.yaml:48]: What is the reason for this change?

File: ray-operator/config/samples/ray-service.autoscaler.yaml [hunk 7]
```diff
@@ -111,5 +95,5 @@ spec:
                     cpu: "1"
                     memory: "2Gi"
                   requests:
-                    cpu: "1"
+                    cpu: "500m"
```
Comment (kevin85421) [inline ray-operator/config/samples/ray-service.autoscaler.yaml:98]: It is better to keep the resource request and limit the same. In my understanding, KubeRay will use the container's resource limits to set `rayStartParams`.

File: tests/framework/prototype.py [hunk 8]
```diff
@@ -223,27 +245,82 @@ def assert_rule(self, custom_resource=None, cr_namespace='default'):
...
-        # Create a pod for running curl command, because the service is not exposed.
-        shell_subprocess_run(f"kubectl run curl --image=radial/busyboxplus:curl -n {cr_namespace} "
-            "--command -- /bin/sh -c \"while true; do sleep 10;done\"")
-        success_create = False
-        k8s_v1_api = K8S_CLUSTER_MANAGER.k8s_client_dict[CONST.K8S_V1_CLIENT_KEY]
-        for _ in range(30):
-            resp = k8s_v1_api.read_namespaced_pod(name="curl", namespace=cr_namespace)
-            if resp.status.phase != 'Pending':
-                success_create = True
+    """Using curl to access the deployed application(s) on Ray service"""
```
Comment (kevin85421) [inline tests/framework/prototype.py:248]:
```suggestion
    """Using curl to access the deployed application(s) on RayService"""
```

File: tests/framework/prototype.py [hunk 9]
```diff
@@ -223,27 +245,82 @@ def assert_rule(self, custom_resource=None, cr_namespace='default'):
...
+    )
+
+    def __init__(self, queries: List[Dict[str, str]], start_in_background: bool = False):
+        self.queries = queries
+        self.start_in_background = start_in_background
+
+    def assert_rule(self, custom_resource, cr_namespace):
+        # If curl pod doesn't exist, create one
+        if get_pod("default", "run=curl") is None:
+            start_curl_pod("curl", "default", timeout_s=30)
```
Comment (kevin85421) [inline tests/framework/prototype.py:262]:
```suggestion
            start_curl_pod("curl", cr_namespace, timeout_s=30)
```

File: tests/framework/utils.py [hunk 10]
```diff
@@ -197,6 +217,22 @@ def pod_exec_command(pod_name, namespace, exec_command, check = True):
...
+        f"kubectl run {name} --image=radial/busyboxplus:curl -n {namespace} "
+        "--command -- /bin/sh -c \"while true; do sleep 10;done\""
+    )
+
+    # Wait for curl pod to be created
+    k8s_v1_api = K8S_CLUSTER_MANAGER.k8s_client_dict[CONST.K8S_V1_CLIENT_KEY]
+    start_time = time.time()
+    while time.time() - start_time < timeout_s or timeout_s < 0:
+        resp = k8s_v1_api.read_namespaced_pod(name=name, namespace=namespace)
+        if resp.status.phase != 'Pending':
```
Comment (kevin85421) [inline tests/framework/utils.py:231]: ?
```suggestion
        if resp.status.phase == 'Running':
```

**Review by kevin85421 — APPROVED** [review]
Leave some comments; the rest looks good to me! Having these tests would be very helpful!

File: ray-operator/config/samples/ray-service.autoscaler.yaml (hunk 6 above)
Comment (zcin) [reply ray-operator/config/samples/ray-service.autoscaler.yaml:48]: I didn't make the changes from the original autoscaler file, instead I added autoscaling configurations to the original `config/samples/ray_v1alpha1_rayservice.yaml` file.

**Review by sihanwang41 — APPROVED** [review]

File: ray-operator/config/samples/ray-service.autoscaler.yaml [hunk 11]
```diff
@@ -9,72 +9,56 @@ metadata:
...
-      - name: FruitMarket
-        numReplicas: 1
-        rayActorOptions:
-          numCpus: 0.1
-      - name: DAGDriver
-        numReplicas: 1
-        routePrefix: "/"
-        rayActorOptions:
-          numCpus: 0.1
+  serveConfigV2: |
```
[the remark this answers is not in the document]
Comment (zcin) [reply ray-operator/config/samples/ray-service.autoscaler.yaml:12]: 0.5 per replica, it should scale to 14 replicas in the test so that would require 7 CPUs.

File: ray-operator/config/samples/ray-service.autoscaler.yaml [hunk 12]
```diff
@@ -9,72 +9,56 @@ metadata:
...
+        route_prefix: /signal
+        runtime_env:
+          working_dir: "https://github.com/ray-project/serve_workloads/archive/7a0daafdb86b55118534b04d122b1b79c9ae99f5.zip"
   rayClusterConfig:
     rayVersion: '2.5.0' # should match the Ray version in the image of the containers
     ## raycluster autoscaling config
     enableInTreeAutoscaling: true
     autoscalerOptions:
       upscalingMode: Default
-      resources:
```
[the remark this answers is not in the document]
Comment (zcin) [reply ray-operator/config/samples/ray-service.autoscaler.yaml]: reverted

File: ray-operator/config/samples/ray-service.autoscaler.yaml (hunk 7 above)
Comment (zcin) [reply ray-operator/config/samples/ray-service.autoscaler.yaml]: Changed back!

File: tests/framework/utils.py (hunk 10 above)
Comment (zcin) [reply tests/framework/utils.py:231]: Updated

File: tests/test_sample_rayservice_yamls.py [hunk 13]
```diff
@@ -1,55 +1,338 @@
...
+            for cr_event in cr_events:
+                cr_event.trigger()
+
+    def test_service_autoscaling(self, set_up_cluster):
+        filename = "ray-service.autoscaler.yaml"
+        path = CONST.REPO_ROOT.joinpath("ray-operator/config/samples/").joinpath(filename)
+
+        scale_up_rule = AutoscaleRule(
+            query={"path": "/", "json_args": {}},
+            num_repeat=20,
```
Comment (kevin85421) [inline tests/test_sample_rayservice_yamls.py:310]: Does it mean sending 20 requests? How do we know it will scale up to 14 replicas (7 CPUs) instead of another number, such as 5 replicas?

Comment (kevin85421) [reply tests/test_sample_rayservice_yamls.py:310]: I do not understand when will the Serving autoscaling be triggered.

Comment (zcin) [reply tests/test_sample_rayservice_yamls.py:310]: I use the deployment from https://github.com/ray-project/serve_workloads/blob/main/autoscaling_test/blocked.py. If you send requests to this deployment, the request will block indefinitely until https://github.com/ray-project/serve_workloads/blob/main/autoscaling_test/signaling.py releases the "lock". The target number of requests ongoing requests per replica is 1, so serve will try to add enough replicas to serve all 20 requests (since all requests are blocked).

Comment (kevin85421) [reply tests/test_sample_rayservice_yamls.py:310]: Thank you for the explanation! We may need to add comments to both this test and the YAML file.

**Review by kevin85421 — APPROVED** [review]
LGTM

File: tests/test_sample_rayservice_yamls.py (hunk 13 above)
Comment (zcin) [reply tests/test_sample_rayservice_yamls.py:323]: Good point. I've added comments to both the test and yaml.
